python/cudf/cudf/_lib/parquet.pyx

```diff
@@ -174,25 +175,36 @@ cpdef read_parquet(filepaths_or_buffers, columns=None, row_groups=None,
 
     column_names = [x.decode() for x in c_out_table.metadata.column_names]
 
-    # Access the Parquet user_data json to find the index
+    # Access the Parquet per_file_user_data to find the index
     index_col = None
     is_range_index = False
-    cdef map[string, string] user_data = c_out_table.metadata.user_data
-    json_str = user_data[b'pandas'].decode('utf-8')
-    meta = None
-    if json_str != "":
-        meta = json.loads(json_str)
-        if 'index_columns' in meta and len(meta['index_columns']) > 0:
-            index_col = meta['index_columns']
-            if isinstance(index_col[0], dict) and \
-                    index_col[0]['kind'] == 'range':
-                is_range_index = True
-            else:
-                index_col_names = OrderedDict()
-                for idx_col in index_col:
-                    for c in meta['columns']:
-                        if c['field_name'] == idx_col:
-                            index_col_names[idx_col] = c['name']
+    cdef vector[unordered_map[string, string]] per_file_user_data = \
+        c_out_table.metadata.per_file_user_data
+
+    range_index = []
+    index_col_names = None
+    for single_file in per_file_user_data:
+        json_str = single_file[b'pandas'].decode('utf-8')
+        meta = None
+        if json_str != "":
+            meta = json.loads(json_str)
+            if 'index_columns' in meta and len(meta['index_columns']) > 0:
+                index_col = meta['index_columns']
+
+                if isinstance(index_col[0], dict) and \
+                        index_col[0]['kind'] == 'range':
+                    range_index.append(True)
```

Comment:
There's never a place where `range_index.append(False)` is called, meaning that the later condition `all(range_index)` will always be true if `range_index` is non-empty. That seems like it might be wrong, because in that case you could just track that with a scalar that is initially `True` and set to `False` if any metadata is _not_ a range index.

Reply:
Looks suspicious to me as well. We might benefit from a test for the else branch + multiple files.

Reply:
Addressed this by switching to a single variable